### ISSUE TITLE
Realtime filler cache does not flush until next interval starts

### ISSUE THREAD

**Issue by merose** [yamcs/yamcs#626]
The logic in `RealtimeArchiveFiller.processParameters()` seems to be faulty: cache segments are not flushed, even if they are older than the current telemetry time minus the `sortingThreshold`. In fact, cache segments are never flushed until a new interval starts, so that they are seen as from an earlier interval.

This has two impacts:

1. It's easy to fill up the cache with the default segment size. Instructions below can fill the cache in under three minutes with only a single parameter packet at under 700 packets/sec. (I realize that is a very high rate, but since an interval is 139 minutes long, that means the cache can fill up with parameters updating at only 10Hz.)
2. The parameter archive API stops working once the cache fills up.

## The potentially incorrect logic

Currently, in `RealtimeArchiveFiller.processParameters()`, this logic is used to check for segments to flush (lines 151-164):

~~~
long segStart = segQueue.getStart();
if (t < segStart - pastJumpThreshold) {
    ... new values are very old - warn and flush the cache ...
} else if (t < segQueue.getStart() - sortingThreshold) {
    ... new values are older than we can sort - warn and discard ...
} else {
    segQueue.sendToArchive(segStart - sortingThreshold, pgs -> scheduleWriteToArchive(pgs));
}
~~~

Note that `segStart` on the next-to-last line is the earliest time stored in the cache. `sendToArchive` then looks for segments either from an earlier interval or that end before `segStart - sortingThreshold`. But any segments that end during the current interval will never match, so no cache flushing will occur for segments in the current interval.

### Possible fix

It looks like the cutoff time for checking for segments to write should be `t - sortingInterval` instead:

~~~
segQueue.sendToArchive(t - sortingThreshold, pgs -> scheduleWriteToArchive(pgs));
~~~

I have tried this using a modified RealtimeArchiveFiller and the cache flushes segments after the sortingThreshold expires. With the high packet rate, as mentioned above, the cache is stable and does not fill up.

## To reproduce

1. Clone this modified quickstart repository.
~~~
$ git clone -b cache-not-flushing git@github.com:merose/quickstart.git
~~~
2. Build the project.
~~~
$ mvn package
~~~
3. Start Yamcs.
~~~
$ target/bundle-tmp/bin/yamcsd
~~~
4. In another terminal window, start packet delivery with 1ms delay between packets.
~~~
$ python3 send_packets.py --delay 1
~~~
5.. In another terminal window, display the latest two values of the parameter.
~~~
$ curl 'http://localhost:8090/api/archive/myproject/parameters/myproject/value?limit=2'
{
  "parameter": [{
    "id": {
      "name": "value",
      "namespace": "/myproject"
    },
    "rawValue": {
      "type": "UINT32",
      "uint32Value": 1018
    },
    "engValue": {
      "type": "UINT32",
      "uint32Value": 1018
    },
    "generationTime": "2021-10-29T21:21:20.923Z",
    "acquisitionStatus": "ACQUIRED",
    "generationTimeUTC": "2021-10-29T21:21:20.923Z"
  }, {
    "id": {
      "name": "value",
      "namespace": "/myproject"
    },
    "rawValue": {
      "type": "UINT32",
      "uint32Value": 1019
    },
    "engValue": {
      "type": "UINT32",
      "uint32Value": 1019
    },
    "generationTime": "2021-10-29T21:21:20.921Z",
    "acquisitionStatus": "ACQUIRED",
    "generationTimeUTC": "2021-10-29T21:21:20.921Z"
  }],
  "continuationToken": "eyJ0aW1lIjoxNjM1NTQyNTE3OTIwfQ"
}$
~~~
6. Wait two minutes. `$ sleep 120`
7. Try to retrieve parameter values again, failing this time with no results returned.
~~~
$ curl 'http://localhost:8090/api/archive/myproject/parameters/myproject/value?limit=2'
{
}$
~~~
8. View coincident error messages in the log.
~~~
$ egrep 'WARN|ERROR' target/bundle-tmp/lg/yamcs-server.log.0 | head -3
Oct 29 14:23:08.198 myproject [19] org.yamcs.parameterarchive.RealtimeArchiveFiller [WARNING] Realtime parameter archive queue full.Consider increasing the writerThreads (if CPUs are available) or using a back filler
Oct 29 14:23:08.199 myproject [19] org.yamcs.parameterarchive.RealtimeArchiveFiller [WARNING] Realtime parameter archive queue full.Consider increasing the writerThreads (if CPUs are available) or using a back filler
Oct 29 14:23:08.201 myproject [19] org.yamcs.parameterarchive.RealtimeArchiveFiller [WARNING] Realtime parameter archive queue full.Consider increasing the writerThreads (if CPUs are available) or using a back filler
$
~~~

**Comment by merose** [comment]
The fix for #628 appears to solve this, too. See PR #629.

**Comment by merose** [comment]
Closing as fixed.
